### PR TITLE
COMP: Supress warning in MSVC for non-float complex

### PR DESCRIPTION
**Pull request by blowekamp**
## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

**Comment by N-Dekker** [comment]
Thanks @blowekamp  Can you please add this `_SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING` to the other cxx files where the warnings appear as well? At least, I think Modules\Core\Common\test\itkNumericTraitsTest.cxx and Modules\Core\Common\test\itkMathTest.cxx

**Review by N-Dekker — APPROVED** [review]

**Comment by blowekamp** [comment]
@N-Dekker I was just addressed warning I saw on the SimpleITK dashboard. I was unsure if the definition needed to occur at before the point of template instantiation/usage or before the header include. Are you able to verify if this approach takes care of the warnings in MSVC?

**Comment by dzenanz** [comment]
I changed my `ryzenator` nightly build configuration to also test `ITK_LEGACY_REMOVE:BOOL=OFF`. We should have the results on the dashboard tomorrow.

**Comment by N-Dekker** [comment]
I'm not sure, maybe we should add `_SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING` to the entire project, via CMake. Unless LEGACY_REMOVE=ON, of course.

**Comment by dzenanz** [comment]
I just kicked of a build manually. Hopefully I got the settings right, so this will be tested.

**Comment by blowekamp** [comment]
> I'm not sure, maybe we should add `_SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING` to the entire project, via CMake. Unless LEGACY_REMOVE=ON, of course.

If a project is using these deprecated std::complex types, IHMO they should get the warning as an indication they may get undefined behavior from their compiler. However, when compiling ITK we want to suppress the warning.

IMHO that is the goal here, so that could be a "private" define for the project.

**Comment by dzenanz** [comment]
The build is [finished](https://open.cdash.org/build/8445615). It doesn't have any warnings.

Aaaand I just figured out why. This build is done using VS2017, and VS2022 has introduced the warning.

**Comment by dzenanz** [comment]
I have changed my dashboard build configuration to use multiple versions of Visual Studio. It should be exercised overnight.

**Comment by dzenanz** [comment]
Here are the [warnings](https://open.cdash.org/viewBuildError.php?type=1&buildid=8446756). A sample:
```log
M:\Dashboard\ITK\Modules\Core\Common\src\itkNumericTraits.cxx(76,86): warning C4996: 'std::complex<char>::complex': warning STL4037: The effect of instantiating the template std::complex for any type other than float, double, or long double is unspecified. You can define _SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING to suppress this warning.[M:\Dashboard\ITK-build\Modules\Core\Common\src\ITKCommon.vcxproj]
M:\Dashboard\ITK\Modules\Core\Common\test\itkMathTest.cxx(674,37): warning C4996: 'std::complex<int>::complex': warning STL4037: The effect of instantiating the template std::complex for any type other than float, double, or long double is unspecified. You can define _SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING to suppress this warning. [M:\Dashboard\ITK-build\Modules\Core\Common\test\itkMathTest.vcxproj]
M:\Dashboard\ITK\Modules\Core\Common\test\itkNumericTraitsTest.cxx(1276,43): warning C4996: 'std::complex<char>::complex': warning STL4037: The effect of instantiating the template std::complex for any type other than float, double, or long double is unspecified. You can define _SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING to suppress this warning. [M:\Dashboard\ITK-build\Modules\Core\Common\test\ITKCommon1TestDriver.vcxproj]
M:\Dashboard\ITK\Modules\Core\Common\include\itkNumericTraits.h(1939,16): warning C4996: 'std::complex<long>::complex': warning STL4037: The effect of instantiating the template std::complex for any type other than float, double, or long double is unspecified. You can define _SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING to suppress this warning. [M:\Dashboard\ITK-build\Modules\Core\Common\test\ITKCommon1TestDriver.vcxproj]
```

**Comment by dzenanz** [comment]
This PR will silence many of the warnings. But definitely not the ones coming from `itkMathTest.cxx`, and probably not those coming from `itkNumericTraitsTest.cxx`.

**Comment by blowekamp** [comment]
OK, so the define before template instantiation was successful to suppress the warning. It is preferred to add the define in the code or as private properties for compiler flags?

**Comment by dzenanz** [comment]
I don't have a preference for this.

File: Modules/Core/Common/include/itkMath.h [hunk 1]
```diff
@@ -40,6 +40,11 @@
  */
 #include <vxl_version.h>
 
+#if !defined(ITK_LEGACY_REMOVE)
+// Supress MSVC warnings
+#  define _SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING
```
Comment (dzenanz) [inline Modules/Core/Common/include/itkMath.h:45]: I think this should go into the cxx file, not the header file. Somewhere around here:
https://github.com/InsightSoftwareConsortium/ITK/blob/ea231d600298a16396a31742b5f26675e6c6334c/Modules/Core/Common/test/itkMathTest.cxx#L674

**Review by dzenanz — COMMENTED** [review]
Thanks for the improvement.

**Review by dzenanz — APPROVED** [review]

**Pull request merged**